File: company/alibaba/findKthLargest/findKthLargest.go
package findKthLargest

func findKthLargest(nums []int, k int) int {
	var res int
	lo, hi := 0, len(nums)-1
	for {
		p := partition(nums, lo, hi)
		if p < k-1 {
			lo = p + 1
		} else if p > k-1 {
			hi = p - 1
		} else {
			res = nums[p]
			break
		}
	}
	return res
}

func partition(nums []int, lo, hi int) int {
	pivot, l, r := nums[lo], lo+1, hi
	for l <= r {
		if nums[l] < pivot && nums[r] > pivot {
			nums[l] ,nums[r] = nums[r], nums[l]
			l++
			r--
		}
		if nums[l] >= pivot {
			l++
		}
		if nums[r] <= pivot {
			r--
		}
	}
	nums[lo], nums[r] = nums[r], nums[lo]
	return r
}

File: company/alibaba/findKthLargest/findKthLargest.py
import heapq


# for heap implementation
class Solution:
    def findKthLargest(self, nums, k):
        h = []
        for i in nums:
            if len(h) < k:
                heapq.heappush(h, i)
                continue
            if i > h[0]:
                heapq.heappushpop(h, i)
                continue
        return h[0]


if __name__ == '__main__':
    Solution().findKthLargest([3, 2, 1, 5, 6, 4], 2)
